### DISCUSSION TITLE
Running raw complicated queries with sequelize

### DISCUSSION THREAD

**Discussion by monishgithub** [sequelize/sequelize#14561] [Help & Questions]
I have a peculiar problem. I have a joined query due to the distribution of data in various table. I want to run this as raw sql query using node module for sequelize as alot existing code is associated using sequelize orm. I do not want to use orm here but just a plain raw query. The packages are     "sequelize": "^4.44.4",
    "sequelize-auto": "^0.4.29",
    "sequelize-cli": "^4.1.1",

    SELECT testmodel.bacode, testmodel2.paopkey from (SELECT ptr.paopkey as paopkey, psgkey.pskey from (SELECT * FROM	pts where paopkey = 4 and ptr_active = 1) ptr,
    pt_sg psgkey where ptr.pkey = psgkey.pkey) testmodel2,
    pt_act testmodel where testmodel.pskey =  testmodel2.pskey;

Any advise if the above query can be run as raw query and the data can be captured as result.
Also when I run the below code with below sequelize code, I get 
the error **"_sequelize.default.query is not a function"**. I have also created a connection with sequelize the db and it still shows the same error.  "**_datasource.default.getDBClient.query is not a function**"

    import Sequelize from 'sequelize';
    async function getTestDataList(opcoKey) {
      const testdata = await Sequelize.query(`SELECT testmodel.bacode, testmodel2.paopkey from (SELECT ptr.paopkey as paopkey, psgkey.pskey from (SELECT * FROM	pts where paopkey = 4 and ptr_active = 1) ptr,
      pt_sg psgkey where ptr.pkey = psgkey.pkey) testmodel2,
      pt_act testmodel where testmodel.pskey =  testmodel2.pskey;`, 
      { type: Sequelize.QueryTypes.SELECT }).then(function (results) {
         // SELECT query - use then
      })
      return toPlain(testdata);
    }

**Comment by ephys** [comment]
You're using the Sequelize class. [`query`](https://sequelize.org/api/v7/classes/sequelize#query) is a method on Sequelize instances

**Comment by monishgithub** [comment]
I know that, I even tried it with instance and it still gave it. Pls dont close the ticket without answering the questions. The db client is actually using sequelize instances

**Reply by ephys** [reply]
We close issues that are not bugs, feature requests, or problems in our documentation (as indicated in our issue template). Issues are our backlog for things we need to actively fix.

I've moved this to github discussions

---

I can't help you more without seeing some actual code. The code you provided uses a sequelize class instead of an instance. If you're using an instance, then something is going on with how you export/import your instance.

I don't think it's the cause of your issue but note that Sequelize 4 has been EOL for a few years, Sequelize 6 and 7 are currently the only maintained versions.

**Reply by monishgithub** [reply]
This is the sequelize instance wrapper created for my application
import config from 'config';
import Sequelize from 'sequelize';

class DataSource {
  constructor() {
    this.DBclient = null;
  }
  initializeDB(dbCredentials) {
    const dataSource = config.datasource;
    const options = {
      host: "127.0.0.1",
      dialect: dataSource.dialect,
      port: dataSource.port,
      logging: false,
    };
    const { password } = dbCredentials || dataSource; 
    this.DBclient = new Sequelize(
      dataSource.database,
      "root",
      "root",
      options,
    );
  }
  getDBClient() {
    return this.DBclient;
  }
}
export default new DataSource();

**Reply by ephys** [reply]
If you `console.log(this.DBclient)` inside `getDBClient()`, does it print the instance or is it undefined?

**Reply by monishgithub** [reply]
It prints the instance.. I am using it to use for other sequelize code in the application. Only when I use to run the raw query it gives me issue.

**Reply by ephys** [reply]
I won't be able to help more without seeing more code, but it's doubtful it's a sequelize issue, that method has been there for a long time and is here in V4 https://github.com/sequelize/sequelize/blob/v4/lib/sequelize.js#L451
